### DISCUSSION TITLE
Ebyte SX1276 Lora module

### DISCUSSION THREAD

**Discussion by javrolijk** [timmbogner/Farm-Data-Relay-System#106] [Q&A]
Hi folks,

Could anyone provide a hand. I'm trying to fogure out how the FDRS sensor code would work with an Ebyte SC1276 module connected on an ESP32 MCU. I'm not a programmer but can follow instructions quite well.

The Ebyte SC1276 consists of the following pins: M0, M1, RXD, TXD, AUX, VCC and GND.

The FDRS Lora sensor configuration code is as follows:
![Screenshot_20220909_222046](https://user-images.githubusercontent.com/70583691/189467734-f5fe56eb-3b01-44e6-b17c-a3ba94353b63.jpg)
Is there any modifications that needs to be done for it to work with this particular Lora module?

Below are some pictures of the Ebyte SC1276 as a reference.

In advance many thanks for any info.

![20184241144586038](https://user-images.githubusercontent.com/70583691/189467775-56426577-148a-4756-a616-9ba9ec106e8d.jpg)
![Screenshot_20220909_230759](https://user-images.githubusercontent.com/70583691/189467779-d44fd959-bf12-486a-aff3-d6b5b4750cf0.jpg)

The ESP32 MCU:
![Screenshot_20220910_000117](https://user-images.githubusercontent.com/70583691/189467858-be35fbb8-eead-4046-bff1-5587694850b9.jpg)

**Comment by timmbogner** [comment]
Unfortunately this is another situation like the REYAX module. The module you have would be more properly known as an Ebyte E32 module (no relation to ESP32), although I see that the Amazon listing makes this unclear. While it is built around the SX1276/8, the manufacturer again added an extra layer for addressing, making it incompatible with other LoRa modules. 
Modules that access the SX1276 chip directly via the SPI interface are what will be compatible with FDRS. If the module has MISO, MOSI, CS, and SCK pins then you you are in good shape. If the module has RXD and TXD then you probably have an incompatible device. [The LoRa library's repo has more on this](https://github.com/sandeepmistry/arduino-LoRa#compatible-hardware).

Your RFM95W mentioned previously should work.

**Reply by javrolijk** [reply]
Many thanks for the prompt reponce, Timm.
Which Lora module would you recommend that works with the FDRS system and with the longest range possible?

**Reply by timmbogner** [reply]
I can't really compare the ranges of different boards, but I have two [TTGO LoRa32 boards](https://randomnerdtutorials.com/ttgo-lora32-sx1276-arduino-ide/) that I use for testing. If the range isn't sufficient I would try an antenna with better gain.
Range can be increased by increasing the power and spreading factor (SF) configs as well.

**Reply by javrolijk** [reply]
Timm, what are your thoughts on the Seeed Wio-E5 mini (link below). Would that work with the FDRS system? Seems like the pins are accessing the Lora interface chip directly.

https://www.seeedstudio.com/LoRa-E5-mini-STM32WLE5JC-p-4869.html

**Reply by timmbogner** [reply]
The E5 module has SPI, but I think they are the SPI pins for the attached microcontroller. I do not think that it would be compatible with the LoRa library used in FDRS. I think a better stand alone module would be [this Adafruit one](https://www.adafruit.com/product/3072?hidden=yes&main_page=product_info&part_id=3072).
HOWEVER
I own an E5 and it does have the ability to access the LoRa PHY layer unobstructed, which means I can and will support the E5 module's UART interface sometime soon.

**Reply by timmbogner** [reply]
I just searched though various marketplaces and there aren't as many standalone modules with standard 2.54mm pitch as I expected. I thought there used to be more cheap, generic versions of that Adafruit one.
 If you're up for it, there are some more reasonably priced RFM95W boards. They will be smaller and trickier to wire up, but are the same as come installed on the Adafruit breakout.
There are much better options available with an onboard ESP32, so I'd really encourage you to take that approach.

**Reply by javrolijk** [reply]
I was looking for something that could provide the greatest range possible (~10km) and hence the E5 caught my attention. It also seems more industrialized as well in some sense. On the other hand, it should also work with the FDRS platform. Have you given some thought to modules that integrate 4G communication?

**Reply by timmbogner** [reply]
I am not an expert, but I think generally the differences in range of LoRa modules should be minimal. The transmission power is set by a configuration parameter and the signal is sent right into an antenna, which is where you'll probably find some room for reception improvement.
I will add E5 support at some point soonish... I have a half-finished draft of it filed away on my computer that just needs a little more work.

I also have a TTGO 4G module that I've been needing to get around to supporting. I might get to that even sooner, as it should be simpler.

**Reply by aviateur17** [reply]
The LoRa transmit and receive distance is also affected by spreading factor and bandwidth both of which are not currently adjustable in the project code but could be added.  See the following discussion on the Sandeep Mistry Github page:
https://github.com/sandeepmistry/arduino-LoRa/issues/90

**Reply by timmbogner** [reply]
Spreading factor is an adjustable global parameter (LORA_SF), but bandwidth isn't yet.

**Reply by javrolijk** [reply]
> Spreading factor is an adjustable global parameter (LORA_SF), but bandwidth isn't yet.

Timm, on that note. How would one go about setting the value for the SF? What are the dependencies? Sorry for asking, I'm just learning as I try to climb the curve. 😅

**Reply by timmbogner** [reply]
It's cool! You change it by editing the "fdrs_globals.h" file inside of your "Arduino/libraries/" folder.
I would encourage you to start using the latest FDRS version too, which is found in the 'dev' branch of the repo. It will become the new official release sometime in the next couple weeks.

**Reply by javrolijk** [reply]
> It's cool! You change it by editing the "fdrs_globals.h" file inside of your "Arduino/libraries/" folder. I would encourage you to start using the latest FDRS version too, which is found in the 'dev' branch of the repo. It will become the new official release sometime in the next couple weeks.

Hi Timm,

Quick question. Should I replace the fdrs_functions.h and all other functions which was downloaded from the git 'main' branch with the functions in the scr folder of the git 'dev' branch?

**Reply by javrolijk** [reply]
I have some code to display sensor values as well as the ID and Gateway mac on these Lilygo Lora modules. Not sure if support for an OLED code can also be integrated in the FDRS system?

![Lilygo-Ttgo-LoRa32-V2-1-1-6-Versie-433-868-915Mhz-ESP32-Lora-Oled-0-96](https://user-images.githubusercontent.com/70583691/190934839-f114f4a8-2b08-4383-b8ac-340363ddba70.jpg)

**Reply by timmbogner** [reply]
Hey! Sorry for the delay, I took a quick hiatus. How are things going?

Hopefully you got your upgrade situation figured out. Replacing them might work, but you're better off installing the repo in the Arduino 'libraries' folder. 

I definitely like the idea of supporting OLEDs! Perhaps the top strip (yellow on some screens) could display the ID and GTWY_MAC, then the bulk of the screen could output messages from the DBG() function. Super-cool!
If you implement it, please do submit it as a pull request, or if you aren't so comfortable with that you could email the code to me.

**Reply by javrolijk** [reply]
Hi Timm,

Thanks for the follow-up.
Indeed, I've implemented the script for the OLED and it seems to work fine. Nonetheless, since I'm not a coder, I've just implemented the code as it seems logical to me based on other sources.

If you like, I can send the script of how I have the 'FDRS Sensor' working and you could review/revise the OLED code in the proper way.

Below is the 'FDRS Lora Sensor' displaying the relevant data on the OLED.
What's your email address?

Br,
Jairo
![VideoCapture_20220928-122415](https://user-images.githubusercontent.com/70583691/192836773-04516f98-26f4-4eed-b17c-0a131de03d79.jpg)

The Lora Gateway displaying the mac address and a packet counter.
![VideoCapture_20220928-123621](https://user-images.githubusercontent.com/70583691/192838371-5baf67ca-8175-47de-9c20-f9c2486780a8.jpg)

**Comment by timmbogner** [comment]
Ay nice! Yeah I'd love to take a gander at your code... timmbogner@gmail.com. I'll add you as a contributor if I end up implementing it.

**Reply by javrolijk** [reply]
> Ay nice! Yeah I'd love to take a gander at your code... [timmbogner@gmail.com](mailto:timmbogner@gmail.com). I'll add you as a contributor if I end up implementing it.

Email sent with the OLED script.

**Comment by lamasseriadipolverara** [comment]
for e32 ebyte module there is this library
https://github.com/xreef/LoRa_E32_Series_Library
@timmbogner it's possible to include it or there are some conflicts?

**Reply by timmbogner** [reply]
It wouldn't be impossible, but it's not a great fit. Among other things it would tie up a UART interface, making it especially hard to use on ESP8266.
